langchain/callbacks/flyte_callback.py

```diff
@@ -0,0 +1,381 @@
+            self.table_renderer().to_html(self.pandas.DataFrame.from_dict(resp))
+            + "\n"
+            + self.table_renderer().to_html(
+                self.pandas.DataFrame.from_records(all_complexity_metrics)
+            ),
+        )
+
+    def on_llm_error(
+        self, error: Union[Exception, KeyboardInterrupt], **kwargs: Any
+    ) -> None:
+        """Run when LLM errors."""
+        self.metrics["step"] += 1
+        self.metrics["errors"] += 1
+
+    def on_chain_start(
+        self, serialized: Dict[str, Any], inputs: Dict[str, Any], **kwargs: Any
+    ) -> None:
+        """Run when chain starts running."""
+        from flytekit import Deck
+        from flytekitplugins.deck.renderer import TableRenderer
+        import pandas as pd
+
+        self.metrics["step"] += 1
+        self.metrics["chain_starts"] += 1
+        self.metrics["starts"] += 1
+
+        resp: Dict[str, Any] = {}
+        resp.update({"action": "on_chain_start"})
+        resp.update(flatten_dict(serialized))
+        resp.update(self.metrics)
+
+        chain_input = ",".join([f"{k}={v}" for k, v in inputs.items()])
+        input_resp = deepcopy(resp)
+        input_resp["inputs"] = chain_input
+
+        Deck(
+            "LLM Chain Start",
+            TableRenderer().to_html(pd.DataFrame.from_dict(input_resp)),
+        )
+
+    def on_chain_end(self, outputs: Dict[str, Any], **kwargs: Any) -> None:
+        """Run when chain ends running."""
+        from flytekit import Deck
+        from flytekitplugins.deck.renderer import TableRenderer
+        import pandas as pd
+
+        self.metrics["step"] += 1
+        self.metrics["chain_ends"] += 1
+        self.metrics["ends"] += 1
+
+        resp: Dict[str, Any] = {}
+        chain_output = ",".join([f"{k}={v}" for k, v in outputs.items()])
+        resp.update({"action": "on_chain_end", "outputs": chain_output})
+        resp.update(self.metrics)
+
+        Deck(
+            "LLM Chain End",
+            TableRenderer().to_html(pd.DataFrame.from_dict(resp)),
+        )
+
+    def on_chain_error(
+        self, error: Union[Exception, KeyboardInterrupt], **kwargs: Any
+    ) -> None:
+        """Run when chain errors."""
+        self.metrics["step"] += 1
+        self.metrics["errors"] += 1
+
+    def on_tool_start(
+        self, serialized: Dict[str, Any], input_str: str, **kwargs: Any
+    ) -> None:
+        """Run when tool starts running."""
+        from flytekit import Deck
+        from flytekitplugins.deck.renderer import TableRenderer
+        import pandas as pd
+
+        self.metrics["step"] += 1
+        self.metrics["tool_starts"] += 1
+        self.metrics["starts"] += 1
+
+        resp: Dict[str, Any] = {}
+        resp.update({"action": "on_tool_start", "input_str": input_str})
+        resp.update(flatten_dict(serialized))
+        resp.update(self.metrics)
+
+        Deck(
+            "LLM Tool Start",
+            TableRenderer().to_html(pd.DataFrame.from_dict(resp)),
```

Comment:
nit: we import pandas , deck, and teblerenderer as `self.pandas`, etc. elsewhere then import again here. Should be consistent. If we're storing as an instance attribute, just call that